static/app/utils/performance/contexts/genericQueryBatcher.tsx

```diff
@@ -0,0 +1,240 @@
+};
+
+const [GenericQueryBatcherProvider, _useGenericQueryBatcher] =
+  createDefinedContext<QueryBatch>({
+    name: 'GenericQueryBatcherContext',
+  });
+
+function mergeKey(query: BatchQueryDefinition) {
+  return `${query.batchProperty}.${query.path}`;
+}
+
+type MergeMap = Record<string, BatchQueryDefinition[]>;
+
+// Builds a map that will contain an array of query definitions by mergeable key (using batch property and path)
+function queriesToMap(collectedQueries: Record<symbol, BatchQueryDefinition>) {
+  const keys = Reflect.ownKeys(collectedQueries);
+  if (!keys.length) {
+    return false;
+  }
+  const mergeMap: MergeMap = {};
+
+  keys.forEach(async key => {
+    const query = collectedQueries[key];
+    mergeMap[mergeKey(query)] = mergeMap[mergeKey(query)] || [];
+    mergeMap[mergeKey(query)].push(query);
+    delete collectedQueries[key];
+  });
+
+  return mergeMap;
+}
+
+function requestFunction(api: Client, path: string, queryObject: QueryObject) {
+  return api.requestPromise(path, queryObject);
+}
+
+function _handleUnmergeableQuery(queryDefinition: BatchQueryDefinition) {
+  const result = requestFunction(
+    queryDefinition.api,
+    queryDefinition.path,
+    queryDefinition.requestQueryObject
+  );
+  queryDefinition.resolve(result);
+}
+
+function _handleUnmergeableQueries(mergeMap: MergeMap) {
+  Object.keys(mergeMap).forEach(async k => {
+    // Using async forEach to ensure calls start in parallel.
+    const mergeList = mergeMap[k];
+
+    if (mergeList.length !== 1) {
+      return;
+    }
+
+    const [queryDefinition] = mergeList;
+    _handleUnmergeableQuery(queryDefinition);
+  });
+}
+
+function _handleMergeableQueries(mergeMap: MergeMap) {
+  Object.keys(mergeMap).forEach(async k => {
+    const mergeList = mergeMap[k];
+
+    if (mergeList.length <= 1) {
+      return;
+    }
+
+    const [exampleDefinition] = mergeList;
+    const batchProperty = exampleDefinition.batchProperty;
+    const requestQueryObject = {...exampleDefinition.requestQueryObject};
+
+    const batchValues = mergeList.map(q => {
+      const batchFieldValue = q.requestQueryObject.query[batchProperty];
+      if (Array.isArray(batchFieldValue)) {
+        return batchFieldValue[0];
+      }
+      return batchFieldValue;
+    });
+
+    requestQueryObject.query[batchProperty] = batchValues;
+
+    const requestPromise = requestFunction(
+      exampleDefinition.api,
+      exampleDefinition.path,
+      requestQueryObject
+    );
+
+    try {
+      const result = await requestPromise;
+      // Unmerge back into individual results
+      mergeList.forEach(queryDefinition => {
+        const propertyName = Array.isArray(
+          queryDefinition.requestQueryObject.query[queryDefinition.batchProperty]
+        )
+          ? queryDefinition.requestQueryObject.query[queryDefinition.batchProperty][0]
+          : queryDefinition.requestQueryObject.query[queryDefinition.batchProperty];
```

Comment:
Does this allow multiple y axis from each query?

Reply:
No, and it currently shouldn't.